internal/controller/controllers/local_cluster_import_controller.go

```diff
@@ -0,0 +1,517 @@
+		return errors.Wrapf(err, "failed to delete AgentClusterInstall %s in namespace %s", name, namespace)
+	}
+	return nil
+}
+
+func (r *LocalClusterImportReconciler) hasLocalManagedCluster(ctx context.Context) (bool, error) {
+	managedCluster := &clusterv1.ManagedCluster{}
+	namespacedName := types.NamespacedName{
+		Name: r.localClusterName,
+	}
+	err := r.client.Get(ctx, namespacedName, managedCluster)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (r *LocalClusterImportReconciler) deleteLocalClusterImportCRs(ctx context.Context) error {
+	err := r.deleteClusterDeployment(ctx, r.localClusterName, r.localClusterName)
+	if err != nil {
+		r.log.Errorf("could not delete local ClusterDeployment due to error %s", err.Error())
+		return err
+	}
+	err = r.deleteAgentClusterInstall(ctx, r.localClusterName, r.localClusterName)
+	if err != nil {
+		r.log.Errorf("could not delete local AgentClusterInstall due to error %s", err.Error())
+		return err
+	}
+	return nil
+}
+
+func (r *LocalClusterImportReconciler) createOrUpdateClusterImageSet(ctx context.Context, release_image string, instance *aiv1beta1.AgentServiceConfig) error {
+	clusterImageSet := hivev1.ClusterImageSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: localClusterImageSetName,
+		},
+	}
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(instance, &clusterImageSet, r.client.Scheme()); err != nil {
+			return err
+		}
+		clusterImageSet.Spec.ReleaseImage = release_image
+		return nil
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, &clusterImageSet, mutateFn)
+	if err != nil {
+		return errors.Wrap(err, "could not create cluster image set")
+	}
+	return nil
+}
+
+func (r *LocalClusterImportReconciler) createOrUpdateNamespace(ctx context.Context, name string, instance *aiv1beta1.AgentServiceConfig) error {
+	namespace := hivev1.ClusterImageSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+	}
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(instance, &namespace, r.client.Scheme()); err != nil {
+			return err
+		}
+		return nil
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, &namespace, mutateFn)
+	if err != nil {
+		return errors.Wrap(err, "could not create or update namespace")
+	}
+	return nil
+}
+
+func (r *LocalClusterImportReconciler) createOrUpdateSecret(ctx context.Context, namespace string, name string, data map[string][]byte, instance *aiv1beta1.AgentServiceConfig) error {
+	secret := v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(instance, &secret, r.client.Scheme()); err != nil {
+			return err
+		}
+		secret.Data = data
+		return nil
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, &secret, mutateFn)
```

Comment:
It might be good to log the operation for all of these if something is actually done.

Reply:
Maybe something like `"%s secret %s/%s", createOrUpdateResult, secret.namespace, secret.name`